src/ucs/datastruct/ptr_array.h

```diff
@@ -120,13 +149,173 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+ * @param [in] name              The name of the ptr array.
+ *
+ * @return Success or failure.
+ *
+ */
+ucs_status_t
+ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array,
+                          uint32_t init_placeholder,
+                          const char *name);
+
+
+/**
+ * Cleanup the locked array.
+ *
+ * @param [in] locked_ptr_array    Pointer to a locked ptr array.
+ *
+ * @note All values should already be removed from it.
+ */
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array);
+
+
+/**
+ * Insert a pointer to the locked array.
+ *
+ * @param [in]  locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in]  value             Pointer to insert. Must be 8-byte aligned.
+ * @param [out] placeholder       Filled with placeholder value.
+ *
+ * @return The index to which the value was inserted.
+ *
+ * Complexity: amortized O(1)
+ *
+ * @note The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array,
+                                     void *value, uint32_t *placeholder_p);
+
+
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index to remove from.
+ * @param [in] placeholder       Value to put in the free slot.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array,
+                                 unsigned index, uint32_t placeholder);
+
+
+/**
+ * Replace pointer in the locked array
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index of slot
+ * @param [in] new_val           Value to put into slot given by index
+ *
+ * @return old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	                           unsigned index, void *new_val);
+
+
+/** Acquire the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array) \
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/** Release the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array) \
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieves a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param [in]  _locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  _index              Index to retrieve the value from.
+ * @param [out] _var                Filled with the value.
+ *
+ * @return Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+#define ucs_ptr_array_locked_lookup(_locked_ptr_array, _index, _var) \
```

Comment:
I think better to remove
- `ucs_ptr_array_locked_lookup()`
- `ucs_ptr_array_locked_for_each()`
they are not thread-safe, but their names can give us a wrong hint about it thread-safety.


or add the real thread-safe variants for them

Reply:
OK, @dmitrygx , @brminich , 

So here's my recommendation - Let's do as follows,
1) Remove ucs_ptr_array_locked_lookup() and ucs_ptr_array_locked_for_each().
2) Keep: ucs_ptr_array_locked_acquire_lock() and ucs_ptr_array_locked_release_lock().
2) Add ucs_ptr_array_locked_super_get() - Then, the user can use the 2 lacquire/lock functions in addition to the ucs_ptr_array_lookup() and ucs_ptr_array_for_each() to iterate through the elements.

@alex--m , @dmitrygx , @brminich , Please comment.

Reply:
> So here's my recommendation - Let's do as follows,
> 
> 1. Remove ucs_ptr_array_locked_lookup() and ucs_ptr_array_locked_for_each().
> 2. Keep: ucs_ptr_array_locked_acquire_lock() and ucs_ptr_array_locked_release_lock().
> 3. Add ucs_ptr_array_locked_super_get() - Then, the user can use the 2 lacquire/lock functions in addition to the ucs_ptr_array_lookup() and ucs_ptr_array_for_each() to iterate through the elements.

👍

Reply:
OK, @dmitrygx  - Please review the latest code.
Following a discussion with @alex--m , please see a nice and elegant solution for most concerns and comment.